File: backend/app/api/smart.py
from typing import Optional
from urllib.parse import urlencode

from fastapi import APIRouter, Body, Depends, Form, HTTPException, Query
from fastapi.responses import RedirectResponse
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.config import settings
from app.core.database import get_db
from app.fhir.server import fhir_server
from app.smart.oauth2 import smart_auth

router = APIRouter(prefix="/smart", tags=["SMART on FHIR"])


@router.get("/.well-known/smart-configuration")
async def smart_configuration():
    base = settings.FHIR_BASE_URL.replace("/fhir", "")
    return {
        "issuer": base,
        "authorization_endpoint": f"{base}/smart/authorize",
        "token_endpoint": f"{base}/smart/token",
        "capabilities": [
            "launch-standalone",
            "launch-ehr",
            "client-public",
            "context-standalone-patient",
            "permission-patient",
            "sso-openid-connect",
        ],
        "scopes_supported": [
            "openid",
            "fhirUser",
            "launch",
            "launch/patient",
            "patient/*.read",
            "offline_access",
        ],
        "response_types_supported": ["code"],
        "code_challenge_methods_supported": ["S256"],
    }


@router.get("/authorize")
async def authorize(
    response_type: str = Query(...),
    client_id: str = Query(...),
    redirect_uri: str = Query(...),
    scope: str = Query(...),
    state: Optional[str] = Query(None),
    code_challenge: Optional[str] = Query(None),
    code_challenge_method: Optional[str] = Query(None),
    launch: Optional[str] = Query(None),
    aud: Optional[str] = Query(None),
    db: AsyncSession = Depends(get_db),
):
    client = await smart_auth.get_client(db, client_id)
    if not client:
        raise HTTPException(400, "Unknown client_id")
    if redirect_uri not in client.redirect_uris:
        raise HTTPException(400, "Invalid redirect_uri")

    # POC: auto-approve. Patient context resolution priority:
    #   1. Launch token created via POST /smart/launch-context (EHR-Launch flow)
    #   2. Fallback to first patient in the store
    patient_id = smart_auth.resolve_launch_context(launch)
    if not patient_id:
        patients = await fhir_server.list_all(db, "Patient")
        patient_id = patients[0]["id"] if patients else None

    scopes = [s for s in scope.split() if s in (client.allowed_scopes or [])]
    code = await smart_auth.create_auth_code(
        db,
        client_id,
        redirect_uri,
        scopes,
        patient_id,
        code_challenge,
        code_challenge_method,
    )

    params: dict = {"code": code}
    if state:
        params["state"] = state
    return RedirectResponse(f"{redirect_uri}?{urlencode(params)}", status_code=302)


@router.post("/launch-context")
async def create_launch_context(
    payload: dict = Body(...),
    db: AsyncSession = Depends(get_db),
):
    """Create a launch token bound to a specific patient.
    Frontend calls this before opening a SMART app so the EHR-Launch flow
    can pre-select the patient context (instead of always defaulting to the
    first patient in the store).
    """
    patient_id = payload.get("patient_id")
    if not patient_id:
        raise HTTPException(400, "patient_id is required")

    # Sanity-check the patient exists in our FHIR store
    if not await fhir_server.read(db, "Patient", patient_id):
        raise HTTPException(404, f"Patient/{patient_id} not found")

    token = smart_auth.create_launch_context(patient_id)
    return {"launch": token, "patient_id": patient_id}


@router.post("/token")
async def token(
    grant_type: str = Form(...),
    code: str = Form(...),
    redirect_uri: str = Form(...),
    client_id: str = Form(...),
    code_verifier: Optional[str] = Form(None),
    db: AsyncSession = Depends(get_db),
):
    if grant_type != "authorization_code":
        raise HTTPException(400, "Unsupported grant_type")

    result = await smart_auth.exchange_code(db, code, client_id, redirect_uri, code_verifier)
    if not result:
        raise HTTPException(400, "Invalid or expired authorization code")
    return result

File: backend-ts/src/api/smart.ts
/**
 * SMART on FHIR OAuth2 endpoints.
 *
 * Port of `backend/app/api/smart.py`. Mounted at `/smart` from main.ts.
 */

import { Hono } from "hono";

import { settings } from "@/core/config";
import { requireSyncApiKey } from "@/core/security";
import { fhirServer } from "@/fhir/server";
import { buildSmartConfiguration, smartAuth } from "@/smart/oauth2";

export const smartApi = new Hono();

smartApi.get("/.well-known/smart-configuration", (c) => {
  const base = settings.FHIR_BASE_URL.replace("/fhir", "");
  return c.json(buildSmartConfiguration(base));
});

smartApi.get("/authorize", async (c) => {
  const responseType = c.req.query("response_type") ?? "";
  const clientId = c.req.query("client_id") ?? "";
  const redirectUri = c.req.query("redirect_uri") ?? "";
  const scope = c.req.query("scope") ?? "";
  const state = c.req.query("state") ?? null;
  const codeChallenge = c.req.query("code_challenge") ?? null;
  const codeChallengeMethod = c.req.query("code_challenge_method") ?? null;
  const launch = c.req.query("launch") ?? null;
  const aud = c.req.query("aud") ?? null;

  // aud is REQUIRED — SMART App Launch obliges clients to send it, and
  // validating it prevents a malicious SMART app from luring the user
  // to authorize against a different server. (audit 2026-06-12 P3:
  // `if (aud)` made the check skippable by simply omitting the param.)
  if (!aud) {
    return c.json(
      {
        error: "invalid_request",
        detail: `aud is required and must equal ${settings.FHIR_BASE_URL}`,
      },
      400,
    );
  }
  const normalize = (u: string) => u.replace(/\/+$/, "");
  if (normalize(aud) !== normalize(settings.FHIR_BASE_URL)) {
    return c.json({ detail: `aud must equal ${settings.FHIR_BASE_URL}` }, 400);
  }

  const client = smartAuth.getClient(clientId);
  if (!client) {
    return c.json({ detail: "Unknown client_id" }, 400);
  }
  const allowedUris = client.redirectUris ?? [];
  if (!allowedUris.includes(redirectUri)) {
    return c.json({ detail: "Invalid redirect_uri" }, 400);
  }

  // Only the authorization-code flow is implemented. RFC 6749 §4.1.2.1:
  // once redirect_uri is validated, an unsupported response_type is
  // reported via error redirect (the client/redirect_uri checks above
  // already direct-400 when we have no trustworthy redirect target).
  if (responseType !== "code") {
    const errParams = new URLSearchParams({ error: "unsupported_response_type" });
    if (state) errParams.set("state", state);
    return c.redirect(`${redirectUri}?${errParams.toString()}`, 302);
  }

  // PKCE mandatory for public clients (SMART App Launch IG §1.0.2).
  if (!client.isConfidential) {
    if (!codeChallenge) {
      return c.json(
        {
          detail:
            "code_challenge is required for public clients (PKCE). See SMART App Launch IG §1.0.2.",
        },
        400,
      );
    }
    // S256 only — discovery advertises S256 only, and 'plain' defeats
    // PKCE's interception protection (audit 2026-06-12 P3).
    if (codeChallengeMethod !== "S256") {
      return c.json(
        { error: "invalid_request", detail: "code_challenge_method must be 'S256'." },
        400,
      );
    }
  }

  // Patient context comes from the EHR-launch handshake — the dashboard
  // mints a launch token via /sync/launch-context (API-key-gated), then
  // hands it to the SMART app, which echoes it back here. Without a
  // valid launch token we refuse rather than silently picking the first
  // patient in the store (that would be a PHI disclosure to any client
  // who can hit this endpoint).
  let patientId = smartAuth.resolveLaunchContext(launch);
  if (!patientId) {
    // Dev/POC escape hatch (audit 2026-06-12 P2-3): auto-picking the
    // first patient in the store is a PHI disclosure, so it requires
    // BOTH keyless mode AND the explicit env opt-in
    // SMART_DEV_AUTOSELECT_PATIENT=1 (read from process.env at request
    // time; intentionally NOT in settings so it stays a dev-only knob).
    const devAutoselect =
      !settings.SYNC_API_KEY && process.env.SMART_DEV_AUTOSELECT_PATIENT === "1";
    if (!devAutoselect) {
      return c.json(
        {
          detail:
            "Standalone launch is disabled. Launch from the dashboard so a launch context is created via /sync/launch-context.",
        },
        400,
      );
    }
    const patients = fhirServer.listAll("Patient");
    patientId = patients.length > 0 ? (patients[0]!.id as string) : null;
  }

  const requestedScopes = scope.split(/\s+/).filter(Boolean);
  const allowed = new Set(client.allowedScopes ?? []);
  const scopes = requestedScopes.filter((s) => allowed.has(s));

  const code = smartAuth.createAuthCode({
    clientId,
    redirectUri,
    scopes,
    patientId,
    codeChallenge,
    codeChallengeMethod,
  });

  const params = new URLSearchParams({ code });
  if (state) params.set("state", state);
  return c.redirect(`${redirectUri}?${params.toString()}`, 302);
});

smartApi.post("/launch-context", requireSyncApiKey, async (c) => {
  let payload: any;
  try {
    payload = await c.req.json();
  } catch {
    return c.json({ detail: "Request body must be JSON" }, 400);
  }
  const patientId = payload?.patient_id;
  if (!patientId) {
    return c.json({ detail: "patient_id is required" }, 400);
  }
  if (!fhirServer.read("Patient", patientId)) {
    return c.json({ detail: `Patient/${patientId} not found` }, 404);
  }
  const token = smartAuth.createLaunchContext(patientId);
  return c.json({ launch: token, patient_id: patientId });
});

smartApi.post("/token", async (c) => {
  // Form-encoded body per OAuth2 spec.
  const form = await c.req.parseBody();
  const grantType = String(form.grant_type ?? "");
  const code = String(form.code ?? "");
  const redirectUri = String(form.redirect_uri ?? "");
  const clientId = String(form.client_id ?? "");
  const codeVerifier = form.code_verifier ? String(form.code_verifier) : null;

  if (grantType !== "authorization_code") {
    return c.json({ detail: "Unsupported grant_type" }, 400);
  }

  const result = smartAuth.exchangeCode({
    code,
    clientId,
    redirectUri,
    codeVerifier,
  });
  if (!result) {
    return c.json({ detail: "Invalid or expired authorization code" }, 400);
  }
  return c.json(result);
});
